docs/architecture/adr-032-typed-events.md

```diff
@@ -0,0 +1,313 @@
+// This should be defined in `types/events.go`
+type EventEmitter func(context.Context, client.Context, ...EventHandler) error
+
+// EventHandler is a type of function that handles events coming out of the event bus
+// This should be defined in `types/events.go`
+type EventHandler func(proto.Message) error
+
+// Sample use of the functions below
+func main() {
+    ctx, cancel := context.WithCancel(context.Background())
+
+    if err := TxEmitter(ctx, client.Context{}.WithNodeURI("tcp://localhost:26657"), SubmitProposalEventHandler); err != nil {
+        cancel()
+        panic(err)
+    }
+
+    return
+}
+
+// SubmitProposalEventHandler is an example of an event handler that prints proposal details
+// when any EventSubmitProposal is emitted. 
+func SubmitProposalEventHandler(ev proto.Message) (err error) {
+    switch event := ev.(type) {
+    // Handle governance proposal events creation events
+    case govtypes.EventSubmitProposal:
+        // Users define business logic here e.g.
+        fmt.Println(ev.FromAddress, ev.ProposalId, ev.Proposal)
+        return nil
+    default:
+        return nil
+    }
+}
+
+// TxEmitter is an example of an event emitter that emits just transaction events. This can and 
+// should be implemented somewhere in the SDK. The SDK can include an EventEmitters for tm.event='Tx' 
+// and/or tm.event='NewBlock' (the new block events may contain typed events) 
+func TxEmitter(ctx context.Context, cliCtx client.Context, ehs ...EventHandler) (err error) {
+    // Instantiate and start tendermint RPC client
+    client, err := cliCtx.GetNode()
+    if err != nil {
+        return err
+    }
+
+    if err = client.Start(); err != nil {
+        return err
+    }
+
+    // Start the pubsub bus
+    bus := pubsub.NewBus()
+    defer bus.Close()
+
+    // Initialize a new error group
+    eg, ctx := errgroup.WithContext(ctx)
+
+    // Publish chain events to the pubsub bus
+    eg.Go(func() error {
+        return PublishChainTxEvents(ctx, client, bus, simapp.ModuleBasics)
+    })
+
+    // Subscribe to the bus events
+    subscriber, err := bus.Subscribe()
+    if err != nil {
+        return err
+    }
+
+	// Handle all the events coming out of the bus
+	eg.Go(func() error {
+        var err error
+        for {
+            select {
+            case <-ctx.Done():
+                return nil
+            case <-subscriber.Done():
+                return nil
+            case ev := <-subscriber.Events():
+                for _, eh := range ehs {
+                    if err = eh(ev); err != nil {
+                        break
+                    }
+                }
+            }
+        }
+        return nil
+	})
+
+	return group.Wait()
+}
+
+// PublishChainTxEvents events using tmclient. Waits on context shutdown signals to exit.
+func PublishChainTxEvents(ctx context.Context, client tmclient.EventsClient, bus pubsub.Bus, mb module.BasicManager) (err error) {
```

Comment:
Where do you get access to the tendermint pubsub bus?
Can you link to where this is wired in app.go or otherwise?

I have been out of core development for a while, but it seems this is not the "recommended" separation of concerns btw. tendermint and the cosmos sdk. But maybe that has changed. And maybe this is better.

Again, I think this is a larger architecture discussion of how tight to couple with the tendermint codebase and internals beyond the abci interface.

Reply:
@ethanfrey the pubsub bus comes from https://github.com/ovrclk/akash/tree/master/pubsub and is a thin wrapper over the websocket `Subscribe` functionality.